include/zephyr/net/net_core.h

```diff
@@ -154,6 +155,42 @@ int net_send_data(struct net_pkt *pkt);
 #define NET_TC_COUNT 0
 #endif /* CONFIG_NET_TC_TX_COUNT && CONFIG_NET_TC_RX_COUNT */
 
+/**
+ * @brief Registration information for a given L3 handler. Note that
+ *        the layer number (L3) just refers to something that is on top
+ *        of L2. So for example IPv6 is L3 and IPv4 is L3, but Ethernet
+ *        based LLDP, gPTP are more in the layer 2.5 but we consider them
+ *        as L3 here for simplicity.
+ */
+struct net_l3_register {
+	/** Store also the name of the L3 type in order to be able to
+	 * print it later.
+	 */
+	const char * const name;
+	/** Handler function for the given protocol type */
+	enum net_verdict (*handler)(struct net_if *iface,
+				    uint16_t ptype,
+				    struct net_pkt *pkt);
+	/** Is L2 address correct */
+	bool (*l2_addr_check)(struct net_if *iface,
+			      uint16_t ptype,
+			      struct net_pkt *pkt,
+			      struct net_linkaddr *dst_lladdr);
+	/** Protocol type */
+	uint16_t ptype;
+};
+
+#define NET_L3_GET_NAME(l3_name) __net_l3_register_##l3_name
+
+#define NET_L3_REGISTER(_name, _ptype, _handler, _is_addr_ok)		 \
+	static const STRUCT_SECTION_ITERABLE(net_l3_register,		 \
+					     NET_L3_GET_NAME(_name)) = { \
+		.ptype = _ptype,					 \
```

Comment:
One optimization could be to sort these on the `ptype` value for faster lookups.

Reply:
Could you elaborate this comment more, I thought that there is not much possibilities to say what order the linker places these.

Reply:
I have something in a downstream project, but realized it's on compatible with ld scripts. I've created a PR https://github.com/zephyrproject-rtos/zephyr/pull/84089 where you can create a named section using a numeric value to sort.

You would be able to use the `ptype` value to sort the sections.

Reply:
Ok, I could add the ptype to the section name but as the value is defined as hex in `ethernet.h` the sections names look like
```
._net_l3_register.static.__net_l3_register_ARP_0x0806_
._net_l3_register.static.__net_l3_register_IPv4_0x0800_
._net_l3_register.static.__net_l3_register_IPv6_0x86dd_
._net_l3_register.static.__net_l3_register_gPTP_0x88f7_
```
Is your PR able to sort these?

Reply:
Sadly not :(